### PR TITLE
Download contest submission source code

### PR DESCRIPTION
**Pull request by hung3a8**
# Description

Type of change: New feature :rocket:

## What

Users can now download all source codes of a contest they have access to.

## Why

Requested by users

# How Has This Been Tested?

- [x] Test locally
- [x] Test with file-only languages

image
image
image

Sample exported file:  [bedao_m07-data.zip](https://github.com/VNOI-Admin/OJ/files/7887246/bedao_m07-data.zip)

# Checklist

- [x] I have explained the purpose of this PR.
- [x] I have performed a self-review of my code
- [ ] I have made corresponding changes to the README/documentation
- [x] Attached screenshots (if applicable).

### CODE REVIEW

File: judge/views/contests.py [hunk 1]
```diff
@@ -1157,3 +1162,61 @@ def post(self, request, *args, **kwargs):
...
+
+        submission_zip = zipfile.ZipFile(output_dir, 'w', zipfile.ZIP_DEFLATED)
+
+        for user in users:
+            username = user.user.user.username
+            user_dir = username
+            user_history_dir = os.path.join(user_dir, '$History')
+
+            problems = set()
+            submissions = user.submissions.order_by('-id') \
```
Comment (leduythuccs) [inline judge/views/contests.py:1196]: `order_by('-points')`

File: judge/views/contests.py [hunk 2]
```diff
@@ -1157,3 +1162,61 @@ def post(self, request, *args, **kwargs):
...
+                .select_related('problem__problem', 'submission__source', 'submission__language') \
+                .values_list('problem__problem__code', 'submission__source__source',
+                             'submission__language__extension', 'submission__id', 'language__file_only')
+
+            for problem, source, ext, sub_id, file_only in submissions:
+                if problem not in problems:  # Last submission
+                    problems.add(problem)
+                    if file_only:
+                        # Get the basename of the source as it is an URL
+                        filename = os.path.basename(source)
```
Comment (leduythuccs) [inline judge/views/contests.py:1206]: The file_only url is structured as: `/submission_file/<contest_id>/<user_id>/<file_name>.<ext>`, so using os.path.basename will only get you the `<file_name>.<ext>` part (example: https://oj.vnoi.info/src/611645)

File: judge/views/contests.py [hunk 3]
```diff
@@ -1157,3 +1162,61 @@ def post(self, request, *args, **kwargs):
...
+                        submission_zip.write(
+                            default_storage.open(os.path.join(settings.SUBMISSION_FILE_UPLOAD_MEDIA_DIR, filename)),
+                            os.path.join(user_dir, f'{problem}.{ext}'),
+                        )
+                        pass
+                    else:
+                        submission_zip.writestr(os.path.join(user_dir, f'{problem}.{ext}'), source)
+                else:
+                    if file_only:
+                        pass
```
Comment (leduythuccs) [inline judge/views/contests.py:1216]: add comment: Don't store submission history for file_only languages

File: judge/views/contests.py [hunk 4]
```diff
@@ -1157,3 +1162,61 @@ def post(self, request, *args, **kwargs):
...
+
+            problems = set()
+            submissions = user.submissions.order_by('-id') \
+                .select_related('problem__problem', 'submission__source', 'submission__language') \
+                .values_list('problem__problem__code', 'submission__source__source',
+                             'submission__language__extension', 'submission__id', 'language__file_only')
+
+            for problem, source, ext, sub_id, file_only in submissions:
+                if problem not in problems:  # Last submission
+                    problems.add(problem)
```
Comment (leduythuccs) [inline judge/views/contests.py:1203]:
```python
if problem not in problems: 
    # save the submission has the highest points
    problems.add(problem)
```

File: judge/views/contests.py [hunk 5]
```diff
@@ -1157,3 +1162,61 @@ def post(self, request, *args, **kwargs):
...
+        contest = self.get_object()
+        output_dir = self.export_contest_submissions(contest)
+        response = HttpResponse(File(open(output_dir, 'rb')),
+                                content_type='application/zip')
+        os.remove(output_dir)
+        response['Content-Disposition'] = 'attachment; filename="%s-submissions.zip"' % (contest.key)
+        return response
+
+    def export_contest_submissions(self, contest):
+        output_dir = tempfile.mktemp(suffix='.zip')
```
Comment (leduythuccs) [inline judge/views/contests.py:1184]: More comments about the main logic of the function (which submission will be stored, how the folder will be structured, ...)

**Review by leduythuccs — CHANGES_REQUESTED** [review]
nice works, small fixs required

**Comment by hieplpvip** [comment]
Will update the documentation after this PR is merged.

**Review by leduythuccs — APPROVED** [review]
LGTM

**Pull request merged**